### PR TITLE
Elasticsearch: fix support for multiple indices

### LINKED ISSUE #71092: Elasticsearch: Add backend support for multiple indexes
Initially reported [here](https://github.com/grafana/support-escalations/issues/6606). When Elastic is in frontend mode it supports a comma-separated list of indexes, but when it's called as a backend data source it does not.

E.g. the data source settings test succeeds.

Image

But then you get no logs in Explore.

Image

As a side effect, this is affecting alerting.

### PR DESCRIPTION
**Pull request by matyax**
Originally reported in an escalation, this PR fixes support for multiple configured indices, as a comma-separated list.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/71092

**Special notes for your reviewer:**

To verify:
- Provision an Elasticsearch instance with multiple indices
- Configure the data source with a comma-separated list of indices

 Expected results: without the fix, no data is returned by the data source; with the code changes, it should work as usual.

### CODE REVIEW

File: pkg/tsdb/elasticsearch/client/client.go [hunk 1]
```diff
@@ -207,7 +207,7 @@ func (c *baseClientImpl) createMultiSearchRequests(searchRequests []*SearchReque
 			header: map[string]interface{}{
 				"search_type":        "query_then_fetch",
 				"ignore_unavailable": true,
-				"index":              c.indices,
```
Comment (svennergr) [inline pkg/tsdb/elasticsearch/client/client.go:210]: Something we should ideally also do is to dedupe the array. With a "daily" pattern and the index as `[logs-]YYYY.MM.DD,[test],[metrics-]YYYY.MM.DD` we would correctly end up in multiple `[logs-]YYYY.MM.DD` and `[metrics-]YYYY.MM.DD` indices, but `test` will also be duplicated, i.e. ending in a total of `logs-2023.07.05,test,metrics-2023.07.05,logs-2023.07.06,test,metrics-2023.07.06,logs-2023.07.07,test,metrics-2023.07.07` for 3 days.

File: pkg/tsdb/elasticsearch/client/client_test.go [hunk 2]
```diff
@@ -210,7 +211,7 @@ func TestClient_Index(t *testing.T) {
 			jHeader, err := simplejson.NewJson(headerBytes)
 			require.NoError(t, err)
 
-			assert.Equal(t, test.indexInRequest, jHeader.Get("index").MustStringArray())
+			assert.Equal(t, strings.Join(test.indexInRequest, ","), jHeader.Get("index").MustString())
```
Comment (svennergr) [inline pkg/tsdb/elasticsearch/client/client_test.go:214]: would be great if we ensure with a test that multiple indices are joined.

File: pkg/tsdb/elasticsearch/client/client.go [hunk 3]
```diff
@@ -207,7 +207,7 @@ func (c *baseClientImpl) createMultiSearchRequests(searchRequests []*SearchReque
...
+				"index":              strings.Join(c.indices, ","),
```
Comment (svennergr) [inline pkg/tsdb/elasticsearch/client/client.go:210]: Even though this seems to work and I might have pointed you in a wrong direction (sorry), I am starting to be a bit unsure if that's the right approach. After reading https://www.elastic.co/guide/en/elasticsearch/reference/current/search-multi-search.html 
```
index
(Optional, string or array of strings) Data streams, indices, and aliases to search. Supports wildcards (*). Specify multiple targets as an array.

If this parameter is not specified, the <target> request path parameter is used as a fallback.
```

We should either use an array here or use the correct `<target>` in the request path parameter. WDYT? We can also discuss with the rest of the squad next week.

**Review by svennergr — COMMENTED** [review]
Left a few comments, maybe we should quickly talk about that on Monday's meeting.

File: pkg/tsdb/elasticsearch/client/client.go (hunk 3 above)
Comment (matyax) [reply pkg/tsdb/elasticsearch/client/client.go:210]: Definitely! Let's discuss this next week.

**Comment by ivanahuckova** [comment]
I haven't head a detailed look, but I suspect this https://github.com/grafana/grafana/pull/67276 broke it. We decided to change comma separated list to array to be consistent with frontend. So maybe we could just revert it.

**Comment by gabor** [comment]
i did some testing on how "frontend-mode" handles this and:
1. for things where there is no "interpolation", for example `logs`, it sends it as a single-string, like:
```json
{
    "index": "logs"
}
```
this also works if you put more indexes comma-separated, like `logs1,logs2`, it becomes:
```json
{
    "index": "logs1,logs2"
}
```

2. if you use the `pattern` feature, for example `daily pattern`, and set the index to be `[logs-]YYYY.MM.DD`, and query for the last 3 days, you get an array-of-strings:
```json
{
    "index": [
        "logs-2023.07.08",
        "logs-2023.07.09",
        "logs-2023.07.10"
    ]
}
```

3. if you have a pattern, and also comma-separated, it does not work, like, if i extend [2] to be `[logs-]YYYY.MM.DD,logs-*`, it will generate a broken index-string

**Comment by gabor** [comment]
after some thinking, i think [3] from above can be considered unsupported. i see two ways to achieve this:

1. make it a single-string: this PR, or just revert the PR @ivanahuckova mentioned above
2. fix the array-of-strings case, that would need a change like this:
```diff
 func (ip *staticIndexPattern) GetIndices(timeRange backend.TimeRange) ([]string, error) {
        if ip.indexName != "" {
-               return []string{ip.indexName}, nil
+               return strings.Split(ip.indexName, ","), nil
        } else {
                return []string{}, nil
        }
```


i think the single-string approach is more robust, WDYT? @ivanahuckova @matyax

**Comment by svennergr** [comment]
> i think the single-string approach is more robust, WDYT? @ivanahuckova @matyax

I am a bit concerned with the statement in the Elastic docshttps://www.elastic.co/guide/en/elasticsearch/reference/current/search-multi-search.html

> Specify multiple targets as an array.

The other fix could be to rely on this statement
>If this parameter is not specified, the <target> request path parameter is used as a fallback.

and remove the `index` from the `headers` and use it from the request path, where multiple indices can be a single string.

**Comment by matyax** [comment]
Thanks everyone

**Pull request closed without merging**